File: backend/app/schemas/contracts.py
from datetime import datetime, timezone
from typing import Literal
from pydantic import BaseModel, Field

SystemMode = Literal["LIVE_AI", "SIMULATION", "DEMO"]
DepthLevel = Literal["NEAR", "MEDIUM", "FAR"]
MotionState = Literal["STATIC", "MOVING"]
Priority = Literal["LOW", "MEDIUM", "HIGH", "CRITICAL"]
Resolution = Literal["COARSE", "MEDIUM", "FINE", "VERY_FINE"]

class DetectedObject(BaseModel):
    id: int
    object_class: str = Field(serialization_alias="class")
    confidence: float
    bbox: tuple[float, float, float, float]
    center: tuple[float, float]
    estimatedDistance: float | None = None
    depthLevel: DepthLevel
    depthConfidence: float | None = None
    velocity: float | None = None
    motionState: MotionState
    direction: str | None = None
    semanticConfidence: float | None = None
    riskScore: float
    priority: Priority
    resolution: Resolution
    factors: dict[str, float]
    reason: str

class SpatialRegion(BaseModel):
    id: str
    x: float
    y: float
    width: float
    height: float
    depth: float | None = None
    priority: float
    resolution: Resolution
    objectId: int | None = None

class Metrics(BaseModel):
    fps: float | None = None
    inferenceLatencyMs: float | None = None
    endToEndLatencyMs: float | None = None
    cpuPercent: float | None = None
    memoryMb: float | None = None
    objectCount: int
    highPriorityRegions: int
    activeCells: int
    source: Literal["MEASURED", "DEMO", "UNAVAILABLE"]

class VirtualGridZone(BaseModel):
    id: str
    label: str
    distanceBand: str
    cellSizeCmEquivalent: int
    cellCount: int

class VirtualGridStats(BaseModel):
    basis: Literal["VISION_DERIVED_VIRTUAL_GRID"] = "VISION_DERIVED_VIRTUAL_GRID"
    fieldWidthMEquivalent: float
    fieldDepthMEquivalent: float
    uniformCellSizeCmEquivalent: int
    uniformCellCount: int
    adaptiveCellCount: int
    cellReductionPercent: float
    modeledUniformStorageKb: float
    modeledAdaptiveStorageKb: float
    modeledStorageReductionPercent: float
    zones: list[VirtualGridZone]

class SystemStatus(BaseModel):
    mode: SystemMode
    camera: Literal["CONNECTED", "UNAVAILABLE", "NOT_REQUESTED"]
    processing: Literal["LOCAL"] = "LOCAL"
    detector: str
    message: str | None = None

class PerceptionSnapshot(BaseModel):
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    objects: list[DetectedObject]
    spatial_regions: list[SpatialRegion]
    metrics: Metrics
    grid_stats: VirtualGridStats
    system_status: SystemStatus

class ConfigUpdate(BaseModel):
    mode: SystemMode | None = None
    cameraIndex: int | None = Field(default=None, ge=0, le=9)

File: frontend/src/types/contracts.ts
export type SystemMode = "LIVE_AI" | "SIMULATION" | "DEMO";
export type DepthLevel = "NEAR" | "MEDIUM" | "FAR";
export type MotionState = "STATIC" | "MOVING";
export type Priority = "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";
export type Resolution = "COARSE" | "MEDIUM" | "FINE" | "VERY_FINE";

export interface DetectedObject {
  id: number;
  class: string;
  confidence: number;
  bbox: [number, number, number, number];
  center: [number, number];
  estimatedDistance?: number;
  depthLevel: DepthLevel;
  depthConfidence?: number;
  velocity?: number;
  motionState: MotionState;
  direction?: string;
  semanticConfidence?: number;
  riskScore: number;
  priority: Priority;
  resolution: Resolution;
  factors: Record<string, number>;
  reason: string;
}

export interface SpatialRegion {
  id: string;
  x: number;
  y: number;
  width: number;
  height: number;
  depth?: number;
  priority: number;
  resolution: Resolution;
  objectId?: number;
}

export interface Metrics {
  fps: number | null;
  inferenceLatencyMs: number | null;
  endToEndLatencyMs: number | null;
  cpuPercent: number | null;
  memoryMb: number | null;
  objectCount: number;
  highPriorityRegions: number;
  activeCells: number;
  source: "MEASURED" | "DEMO" | "UNAVAILABLE";
}

export interface VirtualGridZone {
  id: string;
  label: string;
  distanceBand: string;
  cellSizeCmEquivalent: number;
  cellCount: number;
}

export interface VirtualGridStats {
  basis: "VISION_DERIVED_VIRTUAL_GRID";
  fieldWidthMEquivalent: number;
  fieldDepthMEquivalent: number;
  uniformCellSizeCmEquivalent: number;
  uniformCellCount: number;
  adaptiveCellCount: number;
  cellReductionPercent: number;
  modeledUniformStorageKb: number;
  modeledAdaptiveStorageKb: number;
  modeledStorageReductionPercent: number;
  zones: VirtualGridZone[];
}

export interface PerceptionSnapshot {
  timestamp: string;
  objects: DetectedObject[];
  spatial_regions: SpatialRegion[];
  metrics: Metrics;
  grid_stats: VirtualGridStats;
  system_status: {
    mode: SystemMode;
    camera: "CONNECTED" | "UNAVAILABLE" | "NOT_REQUESTED";
    processing: "LOCAL";
    detector: string;
    message?: string;
  };
}
